Build and deploy auth service in minishift

sbryzak opened the pull request:
Issue: #194 

First step is to add kedge files.

Kedge project can be found here: [](https://github.com/kedgeproject/kedge)

To install (linux):

```
curl -L https://github.com/kedgeproject/kedge/releases/download/v0.5.1/kedge-linux-amd64 -o kedge
chmod +x kedge
sudo mv ./kedge /usr/local/bin/kedge
```

To deploy the kedge files:

1. Make sure minishift is running (minishift start) and you are logged in (oc login).

2. Run the following commands in the fabric8-auth base dir:

```
kedge apply -f kedge/db-auth.yml
kedge apply -f kedge/auth.yml
```

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-1`

sbryzak commented:
I've added a couple of new targets to the Makefile - this one starts minishift, creates a new project and deploys the auth and db images using the kedge files:

```
make dev-auth-openshift
```

Once successfully deployed, the auth service should be available at [](https://minishift.local:31000/api/)

```
make dev-auth-openshift-clean
```

This command should rollback the previous command, cleaning up minishift.

The next step will be to deploy a locally-built image (instead of using the one from the docker registry), and add another target that will only run the db in minishift and run auth on the host.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-2`

sbose78 commented:
Nice!

sbryzak commented:
Added another make target to run only the db in minishift, with the auth server running on the host:

```
make dev-authdb-openshift
```

sbryzak commented:
@pranavgore09 While I had redundant targets called from the dev-auth-openshift target, I now have a new target (dev-authdb-openshift) which runs the auth service on the host, while running just the db in minishift.  It needs to call the other makefile targets (prebuild-check, deps, etc) in order to build.  In light of this, would you still recommend creating a new Makefile just for minishift?  If so how would I link the Makefiles, i.e. be able to call targets in the root Makefile from the minishift Makefile?

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-3`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-4`

pranavgore09 commented:
@sbryzak nope! if we are adding that target then I do not see much value in making a new directory and moving/copying files here and there. IMO we can go ahead with this and in future, if a similar requirement comes in then we can do that part.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-5`

sbryzak commented:
I've added a new make target:

```
make dev-auth-local-openshift
```

This one deploys the db service to local minishift, compiles the sources and builds a new docker image directly in the minishift docker registry, then deploys the new fabric8-auth docker image.  See [this gist](https://gist.github.com/sbryzak/3cde98c9e9833b8147e92f8cdd66b555) for the expected console output for a fresh build (minishift vm deleted, make clean run before build).

I would like to review all of the new target names (dev-auth-openshift, dev-authdb-openshift, dev-auth-local-openshift) as I don't really think they're terribly descriptive - any suggestions here welcome.

The other TODO I still have on this issue is to document all of the new make targets in the developer documentation.  I'll continue working on this today.

alexeykazakov commented on Makefile:
```diff
@@ -296,6 +296,44 @@ $(INSTALL_PREFIX):
+	AUTH_POSTGRES_PASSWORD=mysecretpassword \
+	$(FRESH_BIN)
+
+.PHONY: dev-auth-local-openshift
```
Let's call it dev-openshift

alexeykazakov commented on Makefile:
```diff
@@ -296,6 +296,44 @@ $(INSTALL_PREFIX):
 $(TMP_PATH):
 	mkdir -p $(TMP_PATH)
+	
+.PHONY: dev-auth-openshift
```
Let's call it deploy-auth-openshift

alexeykazakov commented on Makefile:
```diff
@@ -296,6 +296,44 @@ $(INSTALL_PREFIX):
+	AUTH_DEVELOPER_MODE_ENABLED=true \
+	kedge apply -f kedge/db-auth.yml -f kedge/auth.yml
+
+.PHONY: dev-authdb-openshift
```
Let's call it dev-db-openshift

alexeykazakov commented on Makefile:
```diff
@@ -296,6 +296,44 @@ $(INSTALL_PREFIX):
+	-eval `minishift docker-env` && docker login -u developer -p $$(oc whoami -t) $$(minishift openshift registry) && docker build -t fabric8/fabric8-auth:dev bin/docker
+	kedge apply -f kedge/auth-local.yml
+	
+.PHONY: dev-auth-openshift-clean
```
clean-openshift?

sbryzak commented:
Docs updated with endpoint.

sbose78 commented:
I am running into issues with local code being pushed into minishift - have you seen this happen on your side ?

```
[shbose@localhost fabric8-auth]$ make dev-openshift
minishift start --cpus 4
The 'minishift' VM is already running.
./minishift/check_hosts.sh
Hosts entry exists
eval `minishift oc-env` &&  oc login -u developer -p developer && oc new-project auth-openshift
Login successful.

You have access to the following projects and can switch between them with 'oc project <projectname>':

  * auth-openshift
    myproject

Using project "auth-openshift".
Error from server (AlreadyExists): project.project.openshift.io "auth-openshift" already exists
make: [Makefile:326: dev-openshift] Error 1 (ignored)
AUTH_DEVELOPER_MODE_ENABLED=true \
kedge apply -f minishift/kedge/db-auth.yml
service "db-auth" configured
deployment "db-auth" configured
sleep 5s
eval `minishift docker-env` && docker login -u developer -p $(oc whoami -t) $(minishift openshift registry) && docker build -t fabric8/fabric8-auth:dev bin/docker
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Login Succeeded
Sending build context to Docker daemon  19.22MB
Step 1 : FROM fabric8/fabric8-auth:v6a56c54
 ---> 81de9f704f1d
Step 2 : COPY fabric8-auth-linux /usr/local/auth/bin/auth
 ---> Using cache
 ---> 5515c7ff8477
Step 3 : RUN echo chmod +x /usr/local/auth/bin/auth
 ---> Using cache
 ---> 96a5fff053e2
Successfully built 96a5fff053e2
kedge apply -f minishift/kedge/auth-local.yml
service "auth" configured
deployment "auth" configured
```

After this,
the pod fails to come up with the error `Image pull back off`

sbryzak commented:
@sbose78 Could you try listing the docker images in the minishift docker registry?

You can connect to the registry with these commands:
> eval $(minishift docker-env)
> docker login -u developer -p $(oc whoami -t) $(minishift openshift registry)

Then just "docker images" to list the images - you should see something like this in the list:

```
fabric8/fabric8-auth                        dev                 6c715b2d2052        14 hours ago        256MB
```

alexeykazakov commented on minishift/README.md:
```diff
@@ -0,0 +1,76 @@
+
+## Usage
+
+See the developer documentation for make targets to run fabric8-auth on minishift
```
Let's add a link to the dev docs.

alexeykazakov commented on minishift/README.md:
```diff
@@ -0,0 +1,76 @@
+
+##### Install Kedge
+
+Following steps will download and install Kedge on your machine and put it in your $PATH. For more detailed information please visit kedgeproject.org
```
`[kedgeproject.org](http://kedgeproject.org/)`

alexeykazakov commented:
`make clean-openshift ` doesn't do anything for me:

```
make: Nothing to be done for 'clean-openshift'.

```

I still see the auth-openshift project.

sbryzak commented:
@alexeykazakov I think it's more likely that the docker registry service has not started yet. Could you start up minishift before you run the make target, perhaps waiting a few minutes first? You can minishift ssh in to check the processes running (docker ps) for the docker registry. If this is the issue, perhaps I need to put some kind of wait loop in the makefile to wait for the docker registry to start.

alexeykazakov commented:
`make dev-openshift` doesn't re-deploy image if there are changes:
1. make dev-openshift
2. Change something in the source code
3. make dev-openshift
There is no new deployment.

And I don't see any deployment configs. How do I redeploy?

alexeykazakov commented:
> I think it's more likely that the docker registry service has not started yet.

Maybe you are right. Can't reproduce it anymore. Will check the docker registry service if I see it again.

sbryzak commented:
@alexeykazakov My fault for the clean-openshift target not working, I've fixed it now.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-11`

sbryzak commented:
@alexeykazakov I wasn't able to get kedge to perform a proper redeploy, even though the minishift internal docker registry shows that the image has been updated.  As a hack, I've updated the Makefile to simply call kedge delete before calling kedge apply, it's not pretty but it works.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-12`

alexeykazakov commented:
> I wasn't able to get kedge to perform a proper redeploy, even though the minishift internal docker registry shows that the image has been updated. As a hack, I've updated the Makefile to simply call kedge delete before calling kedge apply, it's not pretty but it works.

@pradeepto @LalatenduMohanty could you please take a look and tell us if there is a better way to redeploy?

alexeykazakov commented:
@sbryzak I still think the documentation is not optimal. While we can/should keep the detailed documentation in our Dev/Reference docs we should keep a short version of it in our Readme.

1. Please add Usage of `make dev-minishift` to minishift/readme.md. Similar to https://github.com/pranavgore09/fabric8-wit/blob/1c28230ac10a04dfd57418bae7eeb93f538445a2/minishift/README.md and give the link to our dev doc for details about other targets.
2. Mention minishift in our main readme with a link to minishift.md.

alexeykazakov commented:
@sbryzak it's now fails on a clean minishift:
```
kedge delete -f minishift/kedge/auth-local.yml
Error from server (NotFound): error when stopping "STDIN": services "auth" not found
failed to execute command: failed to execute command: exit status 1
Makefile:324: recipe for target 'dev-openshift' failed
make: *** [dev-openshift] Error 255

```

pranavgore09 commented on Makefile:
```diff
@@ -296,6 +296,46 @@ $(INSTALL_PREFIX):
+.PHONY: clean-openshift
+clean-openshift:
+	-eval `minishift oc-env` &&  oc login -u developer -p developer
+	kedge delete -f minishift/kedge/auth.yml -f minishift/kedge/db-auth.yml
```
IMO you can ignore this as well with `-` prefix

pranavgore09 commented:
@alexeykazakov I am pretty sure that the error you saw 
```
Error from server (NotFound): error when stopping "STDIN": services "auth" not found
failed to execute command: failed to execute command: exit status 1
```
is because there was really no service running at that moment hence delete fails, and we can `ignore` this `exit 1` in makefile safely using `-` as prefix to statement`

pranavgore09 commented on Makefile:
```diff
@@ -296,6 +296,46 @@ $(INSTALL_PREFIX):
+	kedge apply -f minishift/kedge/db-auth.yml
+	sleep 5s
+	-eval `minishift docker-env` && docker login -u developer -p $$(oc whoami -t) $$(minishift openshift registry) && docker build -t fabric8/fabric8-auth:dev bin/docker
+	kedge delete -f minishift/kedge/auth-local.yml
```
you can ignore this as well. otherwise target will fail

sbryzak commented:
I've updated the Makefile to ignore kedge errors (thanks @pranavgore09 ), and updated both the main readme and the minishift readme as per suggestions from @alexeykazakov .  Have also rebased against master so should all be up to date.

alexeykazakov commented:
Works fine now! Thanks @sbryzak and @pranavgore09!
@sbryzak let's add a route to auth-db service in make dev-openshift too so we can access DB via `psql -h minishift.local -U postgres -d postgres -p 31001`
And please also mention how to reach the auth service via minishift.local:31000 and auth-db via `psql -h minishift.local -U postgres -d postgres -p 31001` in minishift/readme.md in the Usage section. Plus add Clean up section too.
And we can merge your PR after that! Great job guys!

sbryzak commented:
@alexeykazakov Usage section has been expanded to describe how to access the service now, plus a cleanup section has been added to the readme.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-14`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-207-15`
